### DISCUSSION TITLE
Defining Javascript classes programmatically in Jint 3

### DISCUSSION THREAD

**Discussion by hcaandersen** [sebastienros/jint#1411] [Q&A]
In Jint 3, how do I either (a) define Javascript classes programmatically or (b) allow scripts to have the keyword, `new`, before function calls?

Scripts like `var c = new MyClass()` worked with the (very) old version that I was using, but fail in Jint 3 with the exception "MyClass is not a constructor".

Note that `MyClass` is currently defined programmatically (in C#) as a function. I've tried to work out how to define classes programmatically, but haven't yet succeeded. I'd prefer not to do so, as my code, which defines them as functions, work fine as long as I don't use the `new` keyword.

**Comment by lahma** [comment]
Did you check [the examples in readme](https://github.com/sebastienros/jint#accessing-net-assemblies-and-classes)?

**Reply by hcaandersen** [reply]
We made a decision early on (10 years ago!) to explicitly define each of the Javascript 'classes' so that we would have complete control over their interfaces. We didn't want to expose CLR objects in Javascript because then (1) they'd have methods like ToString(), Equals() and GetHashCode(), and (2) scripts written by users would have access to other .NET classes, which would post a security risk. We therefore added each 'class' by adding an instance of JsFunction to GlobalScope using the following C# code:

`Visitor.GlobalScope[name] = Visitor.Global.FunctionClass.New(constr);`
where `constr` is a delegate that returns an instance of `JsObject` to which had been added fields, properties and functions.

For example, there's a `File` class (see [here](https://enterprisedt.com/products/completeftp/doc/guide/jssapi/File.html)), that user scripts can instantiate using the `new` keyword, as shown below:

`var myFile = new File("/mydir/myfile.txt");`

Unfortunately, with Jint 3 the `new` gives an error when used with `File` because it's defined as a function rather than a class or constructor. It works if I write

`var myFile = File("/mydir/myfile.txt");`

but this means that the scripts that our users have written will break.

So basically I just want to know how I can either (1) add a non-CLR class explicitly using C# code, or (2) allow the `new` keyword to be used with a regular function.

BTW, we're immensely grateful for all the work you've put into Jint. I sorta feel like we owe something. Will you accept a gift from us?

**Reply by lahma** [reply]
You drive a hard bargain for that `IConstructor` going back to public, and there's an ask already: https://github.com/sebastienros/jint/issues/1144 . I don't find the programming model nice, but if that's what is wanted, maybe it needs to be available then. I'll probably create a PR to publish it and cover with public API tests.

> We didn't want to expose CLR objects in Javascript because then (1) they'd have methods like ToString(), Equals() and GetHashCode(), and (2) scripts written by users would have access to other .NET classes, which would post a security risk. We therefore added each 'class' by adding an instance of JsFunction to GlobalScope using the following C# code:

Interop has gone a long way here, you can create member filters that allow you to have rules what to expose and some have created filters that use their own attributes. like `[ExposeToJs] public void GetOrder(int id) { ... }`.  That's the way I would probably do it, here are some test cases to show usage: https://github.com/sebastienros/jint/blob/main/Jint.Tests/Runtime/InteropTests.MemberAccess.cs . Filters and direct interop expose would probably lead to less code, especially now that Jint has a lot of smarts with method overload discovery and parameter conversion.

**Reply by cyraid** [reply]
@lahma  Any chance JInt could include a MemberAccessJSExposeAttribute that would do the MemberAccess for us? And we could just wire it up? Would save everyone a lot of code from programming it themselves. I see several questions of people asking for this, so it would be beneficial if something similar was included. :)

**Reply by cyraid** [reply]
@lahma sorry for all the questions, but does JInt cache the requests or should I do it myself?

**Reply by lahma** [reply]
Jint offers the flexibility that you can build upon. Someone might use interfaces etc so such attribute would not serve there. Easier for consumers  to build the stuff they need.

I think Jint caches to some extent, you can test around with Jint code base with debugger attached.

**Reply by cyraid** [reply]
@lahma I get that, but it seems a common asked for feature would be what other embedded scripting engines have and that's attributes for methods/fields/etc.

Having those premade classes available so you could easily wire up your own by just using the convenience classes/methods from jint for the common use cases would go a long way.

I get that there may be little bandwidth, but maybe there could be a GitHub issue with a help wanted on it that someone could add? I'm sure it'd be trivial to add some common ones. Then maybe a doc how to wire it up? :)

**Comment by sebastienros** [comment]
@lahma will never get enough gifts https://github.com/sponsors/lahma

**Reply by hcaandersen** [reply]
Sorry, but I've gotten myself confused. Who is the main developer of Jint? Or are there multiple developers who've contributed equally?

**Comment by sebastienros** [comment]
I started the project but getting busy with other projects and work so for some time now @lahma is the boss and deserves all the credits for the recent improvements like ES6 features and performance work. Also why I sponsor him myself ;)

![image](https://user-images.githubusercontent.com/1165805/213594628-2c7f0be0-6973-48e2-b3d6-0e6a1458f047.png)

**Reply by hcaandersen** [reply]
I've given you both gifts. Thanks again!

**Reply by lahma** [reply]
Thank you very much! ❤️
